dali/pipeline/operators/decoder/nvjpeg_decoder_decoupled_api.h

```diff
@@ -0,0 +1,364 @@
+  }
+
+  void ProcessImages(MixedWorkspace* ws) {
+    // Creating output shape and setting the order of images so the largest are processed first
+    // (for load balancing)
+    std::vector<std::pair<size_t, size_t>> image_order(batch_size_);
+    for (int i = 0; i < batch_size_; i++) {
+      image_order[i] = std::make_pair(volume(output_shape_[i]), i);
+    }
+    std::sort(image_order.begin(), image_order.end(),
+              std::greater<std::pair<size_t, size_t>>());
+
+    auto& output = ws->Output<GPUBackend>(0);
+    TypeInfo type = TypeInfo::Create<uint8_t>();
+    output.set_type(type);
+    output.Resize(output_shape_);
+    output.SetLayout(DALI_NHWC);
+
+    CUDA_CALL(cudaEventRecord(master_event_, ws->stream()));
+    for (int i = 0; i < num_threads_; ++i) {
+      CUDA_CALL(cudaStreamWaitEvent(streams_[i], master_event_, 0));
+    }
+
+    for (int idx = 0; idx < batch_size_; ++idx) {
+      const int i = image_order[idx].second;
+
+      const auto &in = ws->Input<CPUBackend>(0, i);
+      const auto file_name = in.GetSourceInfo();
+      cudaStream_t stream = ws->stream();
+      thread_pool_.DoWorkWithID(std::bind(
+            [this, i, file_name, stream, &in, &output](int idx, int tid) {
+              SampleWorker(i, file_name, in.size(), tid,
+                              in.data<uint8_t>(), output.mutable_tensor<uint8_t>(i));
+            }, i, std::placeholders::_1));
+    }
+
+    thread_pool_.WaitForWork();
+    for (int i = 0; i < num_threads_; ++i) {
+      CUDA_CALL(cudaEventRecord(events_[i], streams_[i]));
+      CUDA_CALL(cudaStreamWaitEvent(ws->stream(), events_[i], 0));
+    }
+  }
+
+  // Per sample worker called in a thread of the thread pool.
+  // It decodes the encoded image `input_data` (host mem) into `output_data` (device mem) with
+  // nvJPEG. If nvJPEG can't handle the image, it falls back to DALI's HostDecoder implementation
+  // with libjpeg.
+  void SampleWorker(int sample_idx, string file_name, int in_size, int thread_id,
+                    const uint8_t* input_data, uint8_t* output_data) {
+    EncodedImageInfo& info = output_info_[sample_idx];
+
+    if (!info.nvjpeg_support) {
+      OCVFallback(input_data, in_size, output_data, streams_[thread_id], file_name);
+      return;
+    }
+
+    curr_pinned_buff_[thread_id] = (curr_pinned_buff_[thread_id] + 1) % 2;
+    const int buff_idx = thread_id + (curr_pinned_buff_[thread_id] * num_threads_);
+
+    NVJPEG_CALL(nvjpegStateAttachPinnedBuffer(image_states_[sample_idx],
+                                              pinned_buffer_[buff_idx]));
+    NVJPEG_CALL(nvjpegJpegStreamParse(handle_,
+                                      static_cast<const unsigned char*>(input_data),
+                                      in_size,
+                                      false,
+                                      false,
+                                      jpeg_streams_[buff_idx]));
+
+    CUDA_CALL(cudaEventSynchronize(events_[thread_id]));
```

Comment:
Why we synchronize here?

Reply:
Previous `nvjpegDecodeJpegTransferToDevice` on the same stream/thread could be still working with the pinned buffer

Reply:
OK

Reply:
But now I realise these events_ should be double buffered and follow current pinned_buffer index